### PR TITLE
feat(dtfs2-6966): add audit details to updateGefFacility and refactor

### PR DESCRIPTION
**Pull request by AlexBramhill**
## Introduction :pencil2:
We need to add audit details to portal endpoints

## Resolution :heavy_check_mark:
This PR captures a missed updateGefFacility endpoint

## Miscellaneous :heavy_plus_sign:
- Refactored a fair bit of code in this area to use errors

### CODE REVIEW

File: dtfs-central-api/src/v1/controllers/portal/gef-deal/add-min-activities.controller.js [hunk 1]
```diff
@@ -15,18 +15,19 @@ const { PORTAL_ACTIVITY_LABEL, PORTAL_ACTIVITY_TYPE } = require('../../../../con
...
+  await Promise.all(
+    facilities.map((facility) => {
+      const { _id: facilityId, canResubmitIssuedFacilities } = facility;
+      if (canResubmitIssuedFacilities) {
+        const facilityUpdate = {
+          canResubmitIssuedFacilities: false,
+        };
+        updateFacility({ facilityId, facilityUpdate, auditDetails });
+      }
+      return Promise.resolve();
```
Comment (MarRobSoftwire) [inline dtfs-central-api/src/v1/controllers/portal/gef-deal/add-min-activities.controller.js:28]: I'm not a fan of this but can't think of a good alternative - should we be throwing an error if `!canResubmitIssuedFacilities`? Appreciate this is a big refactor but getting code ick here

Comment (MarRobSoftwire) [inline dtfs-central-api/src/v1/controllers/portal/gef-deal/add-min-activities.controller.js:28]: I'm not a fan of this but can't think of a good alternative - should we be throwing an error if `!canResubmitIssuedFacilities`? Appreciate this is a big refactor but getting code ick here

File: libs/common/src/errors/Invalid-facility-id.error.test.ts
Comment (MarRobSoftwire) [inline libs/common/src/errors/Invalid-facility-id.error.test.ts:1]: File name has a capital (same for deal id)

File: libs/common/src/errors/invalid-deal-id.error.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+export class InvalidDealIdError extends Error {
```
Comment (MarRobSoftwire) [inline libs/common/src/errors/invalid-deal-id.error.ts:1]: Should this extend `ApiError`?

File: dtfs-central-api/src/v1/controllers/portal/gef-facility/update-facility.controller.js [hunk 3]
```diff
@@ -1,62 +1,81 @@
...
-        { _id: { $eq: ObjectId(id) } },
-        { $set: updateBody },
-        { returnNewDocument: true, returnDocument: 'after' },
-      );
+    const facilityUpdateWithAuditRecord = { ...facilityUpdate, auditRecord };
 
-      if (updatedFacility) {
-        // update facilitiesUpdated timestamp in the deal
-        const dealUpdateObj = { facilitiesUpdated: new Date().valueOf() };
+    const updatedFacility = await facilitiesCollection.findOneAndUpdate(
```
Comment (MarRobSoftwire) [inline dtfs-central-api/src/v1/controllers/portal/gef-facility/update-facility.controller.js:30]: If I remember right `findOneAndUpdate` returns a slightly weird & deprecated `ModifyResult` type in mongodb v4. I think the name of this should be something like `updateFacilityResponse`

Comment (MarRobSoftwire) [reply dtfs-central-api/src/v1/controllers/portal/gef-facility/update-facility.controller.js:30]: Using `updateOne` would be preferable but would need further refactoring

File: dtfs-central-api/src/v1/controllers/portal/gef-facility/update-facility.controller.js [hunk 4]
```diff
@@ -1,62 +1,81 @@
...
+    if (error instanceof FacilityNotFoundError) {
+      return res.status(404).send({ status: 404, message: error.message });
+    }
 
-  return res.status(400).send({ status: 400, message: 'Invalid Facility Id' });
+    if (error instanceof InvalidDealIdError || error instanceof InvalidFacilityIdError) {
+      return res.status(400).send({ status: 400, message: error.message });
+    }
+    return res.status(500).send({ status: 500, error });
+  }
```
Comment (MarRobSoftwire) [inline dtfs-central-api/src/v1/controllers/portal/gef-facility/update-facility.controller.js:80]: Ideally could replace this with

```suggestion
    if (error instanceof ApiError) {
      return res.status(error.status).send({ status: error.status, message: error.message });
    }
```

File: libs/common/src/errors/invalid-deal-id.error.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+  constructor(dealId?: string) {
+    const message = dealId ? `Invalid deal ID: ${dealId}` : 'Invalid deal ID';
+    super(message);
```
Comment (MarRobSoftwire) [inline libs/common/src/errors/invalid-deal-id.error.ts:4]: Might've commented twice, but would prefer
```suggestion
import { ApiError } from './api.error';

export class InvalidDealIdError extends ApiError {
  constructor(dealId?: string) {
    const message = dealId ? `Invalid deal ID: ${dealId}` : 'Invalid deal ID';
    super({ message, status: 400 });
```

File: trade-finance-manager-api/src/v1/__mocks__/api.js [hunk 6]
```diff
@@ -82,7 +82,9 @@ module.exports = {
     };
     return Promise.resolve(updatedFacility);
   },
-  updateGefFacility: (facilityId, facilityUpdate) => Promise.resolve(facilityUpdate),
+  updateGefFacility: ({ facilityUpdate }) => {
+    return facilityUpdate;
```
Comment (MarRobSoftwire) [inline trade-finance-manager-api/src/v1/__mocks__/api.js:86]: Should this be async? A badly written function that didn't await `updateGefFacility` could pass a test with this mock right?

File: trade-finance-manager-api/src/v1/api.js [hunk 7]
```diff
@@ -632,7 +632,7 @@ const updateGefFacility = async (facilityId, facilityUpdate) => {
       method: 'put',
       url: `${DTFS_CENTRAL_API_URL}/v1/portal/gef/facilities/${facilityId}`,
       headers: headers.central,
-      data: facilityUpdate,
+      data: { facilityUpdate, auditDetails }, // TODO: DTFS2-6966
```
Comment (MarRobSoftwire) [inline trade-finance-manager-api/src/v1/api.js:635]: If this comment is still needed could you add a bit of detail please :)

File: dtfs-central-api/src/v1/controllers/portal/gef-facility/update-facility.controller.js (hunk 4 above)
Comment (AlexBramhill) [reply dtfs-central-api/src/v1/controllers/portal/gef-facility/update-facility.controller.js:80]: genius

File: dtfs-central-api/src/v1/controllers/portal/gef-deal/add-min-activities.controller.js (hunk 1 above)
Comment (AlexBramhill) [reply dtfs-central-api/src/v1/controllers/portal/gef-deal/add-min-activities.controller.js:28]: I feel this. I think it's a rather large refactor as there has to be understanding as to whether we actually want to fail when the `canResubmitIssuedFacilities` is not truthy. I'm just trying to keep the existing implimentation here, while allowing the code to funciton as it should :)

File: trade-finance-manager-api/src/v1/__mocks__/api.js (hunk 6 above)
Comment (AlexBramhill) [reply trade-finance-manager-api/src/v1/__mocks__/api.js:86]: uopdated

File: libs/common/src/errors/Invalid-facility-id.error.test.ts
Comment (AlexBramhill) [reply libs/common/src/errors/Invalid-facility-id.error.test.ts:1]: updated

File: trade-finance-manager-api/src/v1/api.js (hunk 7 above)
Comment (AlexBramhill) [reply trade-finance-manager-api/src/v1/api.js:635]: removed

**Review by MarRobSoftwire — DISMISSED** [review]

**Review by MarRobSoftwire — APPROVED** [review]

**Pull request merged**